### DISCUSSION TITLE
Any concerns about my asdf-bashly plugin?

### DISCUSSION THREAD

**Discussion by pcrockett** [bashly-framework/bashly#453] [Q&A] [answered]
I created an [asdf](https://asdf-vm.com/) plugin for bashly, and I would like to hear if you have any concerns about it before submitting it to asdf's list of official plugins.

https://github.com/pcrockett/asdf-bashly

Having an `asdf` plugin is nice for two reasons:

* multiple versions of bashly can be installed on a system without knowing anything about Ruby
* the bashly dependency can be specified in the same file as all other `asdf` dependencies ([example here](https://github.com/pcrockett/gramps/blob/main/.tool-versions))

It seems to be working just fine, but I'm not a Ruby developer so I don't know if I have any wrong assumptions that may shoot me in the foot later. Would you mind looking at a few lines of Bash and telling me whether I'm doing things correctly from a Ruby / bashly expert perspective?

Specific pieces I'd like someone to look at:

[This function](https://github.com/pcrockett/asdf-bashly/blob/684bcc05e1cc607ee7714b7105469d357297213b/lib/utils.bash#L35) downloads bashly into a temporary directory by setting `GEM_HOME` and running `gem install bashly:x.x.x`.

[This subshell](https://github.com/pcrockett/asdf-bashly/blob/684bcc05e1cc607ee7714b7105469d357297213b/lib/utils.bash#L57) does two things:

* Moves the downloaded temp directory to the more permanent install directory
* Puts a [bashly wrapper script](https://github.com/pcrockett/asdf-bashly/blob/684bcc05e1cc607ee7714b7105469d357297213b/lib/bashly_wrapper.sh#L9) in a `bin` directory. The wrapper script just runs bashly, but with a different `GEM_HOME` environment variable.

Good? Bad? Don't care? Would love to hear some feedback.

**Comment by DannyBen** [comment]
> Good? Bad? Don't care?

Funny.

Well, first, thanks for this. It is interesting.

A few thoughts in no particular order:

1. At first glance, the two hot zones you pointed out look fine.
2. Personally, I never changed `GEM_HOME` for any reason, so take my previous point with a grain of salt.
3. I see that the plugin does not install ruby, but lists it as a dependency. Shouldn't a package manager install any required dependencies instead of listing them? How is it easier than `gem install bashly` then?
4. If your answer to the previous point is that it allows running multiple bashly versions - why is this needed? And if it is definitely needed, people can use gemfile.
5. The installation script seems complex - is this how it is always done in `asdf`? Seems quite brittle.
6. You say "people don't need to know anything about ruby" when installing with `asdf`, but they don't need to know anything about it when installing with `gem install` either. Or am I missing your point?

**Reply by pcrockett** [reply]
fair. thanks for checking it out. i hate writing huge "quote-and respond" comments on github, but you asked a lot of excellent questions worth answering. so here it goes:

> Shouldn't a package manager install any required dependencies instead of listing them?

well... asdf isn't _exactly_ a package manager... their intro describes it like this: "asdf provides a single interface and configuration file to simplify development workflows, and can be extended to all tools and runtimes via a simple plugin interface."

i could go into why i like it, but i'll spare you :grin:. their [introduction](https://asdf-vm.com/guide/introduction.html) is helpful to wrap your brain around it, and compares it to other similar tools.

> multiple bashly versions - why is this needed?

for the same reason tools like devcontainers, nix, pyenv, devbox, etc exist. making sure all developers are using the same versions of tools, among other reasons.

> And if it is definitely needed, people can use gemfile.

so when i say i know next to nothing about Ruby, i mean: _it's a cool scripting language that runs on the JVM, right? and there's this thing called "rails" that's kinda like django for python?_

if i see an asdf plugin, i'd rather reach for that because i already use it. i don't want to find out what gemfiles are, `gem install`, etc. i'm sure the learning curve is _incredibly_ small, but i'd still rather `apt install ruby` and get on with the asdf plugin.

also, there's a nice benefit to consolidating all my tools in one `.tool-versions` file: i can just run `asdf install` and it'll install the correct version of everything for me: shellcheck, bats, hadolint, etc. and now bashly

> The installation script seems complex

i absolutely agree. i generated the repo from an asdf plugin template. i could have written a simpler plugin from scratch, but (i know this is a recurring theme...) ain't nobody got time for that.

however the code that came with it _just works_ and handles lots of edge cases, so i figured i wouldn't touch it, and just modify those two bits of code i linked to above.

---

whew. anyway. thanks again for the feedback, hopefully that addresses all your questions.

do you have any objections to me adding this to asdf's index of plugins? do you have any more questions?

**Reply by DannyBen** [reply, marked as answer]
> do you have any objections to me adding this to asdf's index of plugins?

Nope.

> apt install ruby

This is usually sufficient, although some distros do not upgrade the Ruby version fast enough. Keep that in mind.

> asdf isn't exactly a package manager

Familiar with it, never liked it for this exact reason. It does not behave like a package manager, and it needs convoluted shell scripts for doing simple tasks (or so I hear...). I have created my own "consolidated" installation tool ([rush](https://github.com/DannyBen/rush-cli)), and installing [all my tools](https://github.com/DannyBen/rush-repo) with it. 

Anyways, thanks for sharing, let me know if I can assist further.

**Reply by pcrockett** [reply]
excellent, thanks. now that i think about it, i used `brew install ruby` so i'm good there.

fwiw _rush_ looks awesome. i def will check that out.

**Reply by pcrockett** [reply]
fwiw, it's now an official asdf plugin: https://github.com/asdf-vm/asdf-plugins/pull/931

also fwiw: https://github.com/pcrockett/rush-repo

**Reply by DannyBen** [reply]
Nice.
